src/TraceEvent/TraceEventSession.cs

```diff
@@ -1368,6 +1368,50 @@ public static unsafe List<string> GetActiveSessionNames()
             return activeTraceNames;
         }
 
+        /// <summary>
+        /// Maximum Number of MaxEtwLoggers the system supports
+        /// </summary>
+        private static int? MaxEtwLoggers = null;
+
+        /// <summary>
+        /// Get the maximum number of ETW loggers supported by the current machine
+        /// </summary>
+        /// <returns>The maximum number of supported ETW loggers</returns>
+        private static int GetETWMaxLoggers()
+        {
+            const string MaxEtwRegistryKey = "SYSTEM\\CurrentControlSet\\Control\\WMI";
+            const string MaxEtwPropertyName = "EtwMaxLoggers";
+            const int DefaultMaxETWLoggers = 64;
+
+            if (MaxEtwLoggers == null)
+            {
+                try
+                {
+                    using (RegistryKey key = Registry.LocalMachine.OpenSubKey(MaxEtwRegistryKey))
+                    {
+                        if (key != null && key.GetValue(MaxEtwPropertyName) != null)
+                        {
+                            if (int.TryParse(key.GetValue(MaxEtwPropertyName).ToString(), out int propertyValue))
+                            {
+                                // Ensure registry was set within permissable range as defined by
+                                // https://docs.microsoft.com/en-us/windows/win32/api/evntrace/nf-evntrace-starttracew
+                                if (propertyValue >= 32 && propertyValue <= 256)
+                                {
+                                    MaxEtwLoggers = propertyValue;
+                                }
+                            }
+                        }
+                    }
+                }
+                catch (Exception) { }
+
+                // If the value does not exist or cannot be read from the registry, return the default value
+                MaxEtwLoggers = MaxEtwLoggers ?? DefaultMaxETWLoggers;
```

Comment:
Can we simplify this and just set `MaxEtwLoggers == 64` by default?  Then you don't need to worry about checking to see if `MaxEtwLoggers` has been set.  Also, you can change the type of `MaxEtwLoggers` to `int` so you don't have to cast it below.

Reply:
Don't we want to check if MaxEtwLoggers has been set?  This was done in order to achieve caching of MaxEtwLoggers.
If we have no value, check and set the value to the registry key or the default if the registry key does not exist.

If we set MaxEtwLoggers to 64 by default, in the instance the registry key does not exist, each time we call the function, we would have no way to know whether this was our first time checking, or if we have already checked and had to set the default value.

Reply:
Good point.  What I was thinking would require another field to keep track of whether or not we've been through the initialization path.  I think what you have now is good.